src/compiler/checker.ts

```diff
@@ -8029,6 +8087,65 @@ namespace ts {
             return links.resolvedType;
         }
 
+        function getActualTypeParameter(type: Type) {
+            return type.flags & TypeFlags.Substitution ? (<SubstitutionType>type).typeParameter : type;
+        }
+
+        function createConditionalType(checkType: Type, extendsType: Type, trueType: Type, falseType: Type, target: ConditionalType, mapper: TypeMapper, aliasSymbol: Symbol, aliasTypeArguments: Type[]) {
+            const type = <ConditionalType>createType(TypeFlags.Conditional);
+            type.checkType = checkType;
+            type.extendsType = extendsType;
+            type.trueType = trueType;
+            type.falseType = falseType;
+            type.target = target;
+            type.mapper = mapper;
+            type.aliasSymbol = aliasSymbol;
+            type.aliasTypeArguments = aliasTypeArguments;
+            return type;
+        }
+
+        function getConditionalType(checkType: Type, extendsType: Type, baseTrueType: Type, baseFalseType: Type, target: ConditionalType, mapper: TypeMapper, aliasSymbol?: Symbol, baseAliasTypeArguments?: Type[]): Type {
+            // Return union of trueType and falseType for any and never since they match anything
+            if (checkType.flags & (TypeFlags.Any | TypeFlags.Never)) {

```

Comment:
I can understand this for `any`, but for `never` I see some issues:

```ts
type T1 = number extends never ? true : false; // false
type T2 = never extends never ? true : false; // boolean
```

This doesn't match with the type system for assignability:
```ts
let a: number, b: never, c: never;
b = a; // error: Type 'number' is not assignable to 'never' (matches T1 above)
b = c; // ok, but fails to match T2 above
```

While these are trivial examples, a more complex example might be:

```ts
type Tuple<A, B = never> = B extends never ? [A] : [A, B];

type T1 = Tuple<number, string>; // [number, string]
type T2 = Tuple<number>; // [number, never]
```

There's no way to hit the `true` case above for `never extends never`. I can understand the behavior for `never extends S` where `S` is any other type, but it seems `never extends never` seems like a very specific case we would want to single out.

Reply:
Yeah, I agree. I'm going to remove the special casing of `never` and just let the regular assignment rules guide the result (i.e. `never` is assignable to `never`, but nothing else is).